synedrion/src/paillier/rsa.rs

```diff
@@ -0,0 +1,282 @@
+            let q_wide = q.into_wide();
+            q.zeroize();
+            q_wide
+        })
+    }
+
+    pub fn p_signed(&self) -> SecretBox<Signed<P::Uint>> {
+        SecretBox::init_with(|| {
+            Signed::new_positive(*self.p().expose_secret(), P::PRIME_BITS).expect("`P::PRIME_BITS` is valid")
+        })
+    }
+
+    pub fn q_signed(&self) -> SecretBox<Signed<P::Uint>> {
+        SecretBox::init_with(|| {
+            Signed::new_positive(*self.q().expose_secret(), P::PRIME_BITS).expect("`P::PRIME_BITS` is valid")
+        })
+    }
+
+    pub fn p_nonzero(&self) -> SecretBox<NonZero<P::Uint>> {
+        SecretBox::init_with(|| NonZero::new(*self.p().expose_secret()).expect("`p` is non-zero"))
+    }
+
+    pub fn q_nonzero(&self) -> SecretBox<NonZero<P::Uint>> {
+        SecretBox::init_with(|| NonZero::new(*self.q().expose_secret()).expect("`q` is non-zero"))
+    }
+
+    pub fn p_wide_signed(&self) -> SecretBox<Signed<P::WideUint>> {
+        SecretBox::init_with(|| self.p_signed().expose_secret().into_wide())
+    }
+
+    pub fn totient(&self) -> &SecretBox<P::Uint> {
+        &self.totient
+    }
+
+    pub fn totient_bounded(&self) -> SecretBox<Bounded<P::Uint>> {
+        SecretBox::init_with(|| {
+            Bounded::new(*self.totient.expose_secret(), P::MODULUS_BITS).expect("`P::MODULUS_BITS` is valid")
+        })
+    }
+
+    pub fn totient_wide_bounded(&self) -> SecretBox<Bounded<P::WideUint>> {
+        SecretBox::init_with(|| self.totient_bounded().expose_secret().into_wide())
+    }
+
+    pub fn totient_nonzero(&self) -> SecretBox<NonZero<P::Uint>> {
+        SecretBox::init_with(|| {
+            NonZero::new(*self.totient.expose_secret()).expect(concat![
+                "φ(n) is never zero for n >= 1; n is strictly greater than 1 ",
+                "because it is (p-1)(q-1) and given that both p and q are prime ",
+                "they are both strictly greater than 1"
+            ])
+        })
+    }
+
+    /// Returns a random in range `[0, \phi(N))`.
+    pub fn random_field_elem(&self, rng: &mut impl CryptoRngCore) -> Bounded<P::Uint> {
+        Bounded::new(
+            P::Uint::random_mod(rng, self.totient_nonzero().expose_secret()),
+            P::MODULUS_BITS,
+        )
+        .expect(concat![
+            "the totient is smaller than the modulus, ",
+            "and thefore can be bounded by 2^MODULUS_BITS"
+        ])
+    }
+}
+
+/// The minimized structure containing the public RSA modulus.
+///
+/// Suitable for serialization or transmission.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub(crate) struct PublicModulusWire<P: PaillierParams>(P::Uint);
+
+impl<P: PaillierParams> PublicModulusWire<P> {
+    fn new(primes: &SecretPrimesWire<P>) -> Self {
+        Self(primes.p.expose_secret().mul_wide(primes.q.expose_secret()))
+    }
+
+    pub fn into_precomputed(self) -> PublicModulus<P> {
+        PublicModulus::new(self)
+    }
```

Comment:
For consistency this should probably be named `into_public_modulus`, or `PublicModulus` should be renamed `PrecomputedModulus`. I think I prefer the latter, because the name `PublicModulus` implies that there is something like a "private modulus", which isn't really the case.